File: 백준 문제 풀이 모음/class 4/11779.py
import sys, heapq
INF = sys.maxsize
input = sys.stdin.readline


n = int(input())
m = int(input())

graph = [[] for i in range(n+1)]


for i in range(m) :
    a, b, c = map(int, input().split())
    graph[a].append((b, c))

start, end = map(int, input().split())
prev_node = [0] * (n+1)
def dijkstra(start) :
    
    distance = [INF] * (n + 1)
    distance[start] = 0
    heap = [(start, 0)]
    #print(distance)

    while heap :
        #print('heap: ', heap)
        now_node, now_cost = heapq.heappop(heap)
        #print('now_node: ' + str(now_node) + ' now_cost: ' + str(now_cost))
        if distance[now_node] < now_cost :
            continue

        for next_node, next_cost in graph[now_node] :
            next_cost = next_cost + now_cost
            #print('next_node: ' + str(next_node) + ' next_cost: ' + str(next_cost))

            if distance[next_node]  > next_cost :
                distance[next_node]  = next_cost
                prev_node[next_node] = now_node
                #print('distance: ', distance)
                #print('prev_node: ', prev_node)
                heapq.heappush(heap, (next_node, next_cost))
    return distance

print(dijkstra(start)[end])

path = [end]
now = end
while now != start :
    now = prev_node[now]
    path.append(now)
path.reverse()
print(len(path))
print(' '.join(map(str, path)))

File: 백준 문제 풀이 모음/kotlin/class 4/11779.kt
import java.util.PriorityQueue

private const val INF = Integer.MAX_VALUE
private lateinit var dist : MutableList<Int>
private lateinit var list : MutableList<MutableList<Node>>
private lateinit var visited : MutableList<Boolean>
private lateinit var back : MutableList<Int>

private class Node (var nodeNum : Int, var weight :Int) : Comparable<Node> {
    override fun compareTo(other: Node): Int {
        return weight - other.weight
    }
}

fun main() = with(System.`in`.bufferedReader()){
    val n = readLine().toInt()
    val m = readLine().toInt()

    dist = MutableList<Int>(n+1) { INF }
    list = MutableList<MutableList<Node>>(n+1) { MutableList<Node>(0) { Node(0,0)} }
    visited =MutableList<Boolean>(n+1) { false }
    back = MutableList<Int> (n+1) { 0 }
    for (i in 0 .. n) {
        back[i] = i
    }

    repeat(m) {
        val input = readLine().split(' ').toList()
        list[input[0].toInt()].add(Node(input[1].toInt(), input[2].toInt()))
    }

    val input = readLine().split(' ').toList()
    val start = input[0].toInt()
    val end = input[1].toInt()

    dijkstra(start, n)

    println(dist[end])

    var path = mutableListOf<Int>()
    path.add(end)

    var now = end
    while (now != start ) {
        now = back[now]
        path.add(now)
    }
    println(path.size)
    for (i in path.size -1 downTo 0) {
        print(path[i])
        print(' ')
    }

}

fun dijkstra(start: Int, n: Int) {
    dist[start] = 0
    val pq = PriorityQueue<Node>()
    pq.offer(Node(start, 0))

    while (pq.isNotEmpty()) {
//        pq.forEach {
//            print(it.nodeNum)
//            print(' ')
//        }
//        println()
        val node = pq.poll()
//        println(node.nodeNum)
        if (!visited[node.nodeNum]) {
            visited[node.nodeNum] = true
            list[node.nodeNum].forEach {
                if (!visited[it.nodeNum] && dist[it.nodeNum] > dist[node.nodeNum] + it.weight) {
                    dist[it.nodeNum] = dist[node.nodeNum] + it.weight
                    pq.offer(Node(it.nodeNum, dist[it.nodeNum]))
                    back[it.nodeNum] = node.nodeNum
                }
            }
        }
    }
}




